libclc/amdgcn-amdhsa/libspirv/misc/sub_group_shuffle.cl

```diff
@@ -37,7 +37,16 @@ __AMDGCN_CLC_SUBGROUP_SUB_I32(unsigned short, t);
 
 #ifdef cl_khr_fp16
 #pragma OPENCL EXTENSION cl_khr_fp16 : enable
-__AMDGCN_CLC_SUBGROUP_SUB_I32(half, DF16_);
+_CLC_DEF half _Z28__spirv_SubgroupShuffleINTELIDF16_ET_S0_j(
+    half Data, unsigned int InvocationId) {
+  union {
+    unsigned i;
+    half h;
+  } tmp;
+  tmp.h = Data;
+  tmp.i = _Z28__spirv_SubgroupShuffleINTELIiET_S0_j(tmp.i, InvocationId);
```

Comment:
Beware that this relies on UB. See https://en.cppreference.com/w/cpp/language/union:

> It is undefined behavior to read from the member of the union that wasn't most recently written.

That said, I don't know of a better way to get it done... Unless we can use some sort of bitcast builtin here. Have you had a look at alternatives?

Reply:
Yeah it is UB according to openclc for at last one half of the 32 bits depending on endianness. 
If openclc inherited memcpy from c99 it would be possible to do it that way cleanly, but it doesn't. It has a section that discusses the issue briefly: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#reinterpreting-data-as-another-type
One alternative that is not UB is to first do a conversion to float, then do as_int(float), and do the reverse after the shuffle. I tried this and compared the result with doing it the way I have here to check that it gives the same result. However doing it that way generates the extra machine instructions to do the float converts etc.
HIP fp16 headers is open source and I can see that they use the union trick. Hence I figured that if they do it we may as well do it. But at the same time I could just do it via the float route.

Reply:
And yeah and I looked at amd clang builtins and I don't think they have a bitcast one (actually cuda doesn't either). For cuda I used inline ptx asm, but for amd this is more dangerous because they don't have an equivalent to ptx that is arch generation/family independent.

Reply:
Ah, fair point. I missed the fact that this is OpenCL C. I think it is still better to not rely on UB, unless the performance overhead is high. The compiler moves fast and it wouldn't be the first time UB suddenly becomes an issue.

Reply:
It makes the add test cases in the reduction test I reenabled in this PR ~10% slower if I go via the float route. I think it is fine to do it like that if you want? I guess that fp16 shuffles are not so likely to be performance critical for much stuff, although I can't say that with confidence.

Reply:
to_X only works for types of the same size. I tried doing a memcpy via copying via char pointers as you suggest but for some reason this doesn't work. I tried a few things and the only way I can do an explicit copy and get the same device code out as via unions is using short* which allows copying a single element: I've made this change in the last commit.

Reply:
> to_X only works for types of the same size.

Right you are. Bummer!

> I tried doing a memcpy via copying via char pointers as you suggest but for some reason this doesn't work.

That is a little concerning. I believe we use that in the device-side implementation of memcpy internally. I wonder if that doesn't work on HIP either. 🤔 

> I tried a few things and the only way I can do an explicit copy and get the same device code out as via unions is copying via short* which allows copying a single element: I've made this change in the last commit.

I believe this is a strict aliasing violation. To my knowledge, it is allowed to do it with char, but not most other types sadly. So if it's between one UB and another, I think I prefer the union solution... Maybe.

Reply:
> > I tried doing a memcpy via copying via char pointers as you suggest but for some reason this doesn't work.
> 
> That is a little concerning. I believe we use that in the device-side implementation of memcpy internally. I wonder if that doesn't work on HIP either. 🤔

I can look into this more.

> 
> I believe this is a strict aliasing violation. To my knowledge, it is allowed to do it with char, but not most other types sadly. So if it's between one UB and another, I think I prefer the union solution... Maybe.

I can instead cast half to unsigned short, and then extend short to int. This works and leads to only one extra instruction compared to the union trick, which is faster than going via float. Since half class uses unsigned short as the storage type this should be safe?

Reply:
> I can instead cast half to unsigned short, and then extend short to int. This works and leads to only one extra instruction compared to the union trick, which is faster than going via float. Since half class uses unsigned short as the storage type this should be safe?

That sounds like a good solution to me!

Reply:
done

Reply:
> Ah, I didn't realize you meant cast the pointer to `unsigned short *`. I still think this is an alias violation, so the compiler could easily go do some magic that gives us the wrong behavior.
> 
I see thanks.
> If `half` has a `unsigned int` storage type behind the scenes, could we maybe use some inline AMD instruction helpers here to do the conversion? It might even be as simple as a no-op or ID operation, assuming the representation is the same in the resulting AMD representation.

Generally I think inline AMD should be avoided since it is like having inline assembly for SASS if it were cuda. For amd I think clang builtins should always be used that are lowered to the appropriate AMD asm.
I've found a non UB way of doing it anyway via `as_ushort`. I had tried this before but using `as_short` etc which didn't work, but the `as_ushort` etc works perfectly.

Reply:
> I've found a non UB way of doing it anyway via `as_ushort`. I had tried this before, but with using `as_short` etc, which didn't work, but the `as_ushort` etc works perfectly.

Actually `as_short` works too, I must have just messed it up the first time.

Reply:
Great, thank you! I like this solution. 😄